_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+      - [Vivisection](#vivisection)
+    - [Soothing light at the end the tunnel](#soothing-light-at-the-end-the-tunnel)
+      - [Enter the SIMD](#enter-the-simd)
+        - [The nail to SIMD's Hammer](#the-nail-to-simds-hammer)
+        - [Soothing light at the end the tunnel](#soothing-light-at-the-end-the-tunnel)
+        - [What's my name?](#whats-my-name)
+- [Part 2 - let's talk about Java](#part-2---lets-talk-about-java)
+  - [Auto Vectorization (AV)](#auto-vectorization-av)
+    - [-- Interlude --](#---interlude---)
+    - [What abouts the Java?](#what-abouts-the-java)
+  - [Explicit Vectorization (EV)](#explicit-vectorization-ev)
+    - [Prelude](#prelude)
+    - [Enter the Dragon](#enter-the-dragon)
+    - [Show me the money](#show-me-the-money)
+      - [Prerequisites](#prerequisites)
+      - [Hello, 世界](#hello-)
+      - [Vanilla](#vanilla)
+      - [With Nuts](#with-nuts)
+        - [Species](#species)
+        - [Read input into SIMD register(s)](#read-input-into-simd-registers)
+        - [Do operation(s) on SIMD register(s)](#do-operations-on-simd-registers)
+        - [Write SIMD register(s) back to input](#write-simd-registers-back-to-input)
+        - [Hello hello](#hello-hello)
+      - [Benchmark Vanilla vs Nuts](#benchmark-vanilla-vs-nuts)
+        - [Let sleeping dogs lie](#let-sleeping-dogs-lie)
+        - [Monkey Wrench Drop - where AV fails and EV thrives](#monkey-wrench-drop---where-av-fails-and-ev-thrives)
+      - [Real world example](#real-world-example)
+        - [Problem statement](#problem-statement)
+        - [Vanilla MinMax() version](#vanilla-minmax-version)
+        - [SIMD MinMax() version](#simd-minmax-version)
+          - [Elephant in the room](#elephant-in-the-room)
+  - [Epilogue](#epilogue)
+- [References](#references)
+
+## Part 1 - background (written by dummies, for curious cats)
+
+### What is _'Vectorization'_?
+
+Without beating about the bush, does Vectorization, explicit or otherwise, have anything to do with the
+fantabulous `java.util.Vector` class? Well, not much.
+
+Allow me to explain with this very long example.
+
+#### Bullshit use-case
+
+Let's say you are writing an image editing tool similar to Photoshop. One of the functions you want to implement is
+adjusting the brightness. In this particular example:
+
+- Your input is an array of 3 bytes of `RGB` values per pixel, meaning a `1920 x 1080` picture would
+  have `3 x 1920 x 1080 = 6220800` bytes.
+- Increasing brightness is basically adding 1 (or more) to each RGB value per pixel
+
+So how would we do that?
+
+```java
+  for (int i = 0; i < 6220800; i++) {
+      rgb[i]++;
+  }
+```
+
+We could probably do something fancy like unrolling the loop (eliminating some bounds checking) to speed this up:
+
+```java
+  for (int i = 0; i < 6220800; i += 2) {
+      rgb[i + 0]++;
+      rgb[i + 1]++;
+  }
+```
+
+The reason we have to do it this way, is because of real world physics. Your fancy shmancy CPU is basically
```

Comment:
The reason we have to do it this way, is because of real world physics. -> what other way? (magic?)

Can't quite process what it is meant to convey :smile: